Updates collapseDims() function and documentation

mruberry opened the pull request:
The collapseDims() function was correctly implemented but it...

(1) made four passes over its inputs
(2) created two new (very tiny) arrays
(3) was very long

This new version collapses the TensorInfo's dimensions with a single pass and only four ints of storage. It is significantly shorter. The function's documentation has also been updated to clarify its effect and return values.

As additional testing, while running the automated test_cuda.py tests the results of this collapseDims() were compared with the version in Master and no differences reported. That is, the changes made by each were the same. This additional testing is important because it is possible that collapseDims() could be rewritten to do nothing and many (all?) tests would still pass.

Automated testing: test_cuda.py (OK), test_nn.py (no regression from master)

mruberry committed:
Updates collapseDims() function and documentation

wickedfoo commented on aten/src/ATen/cuda/detail/TensorInfo.cuh:
```diff
@@ -66,157 +71,70 @@ TensorInfo<T, IndexType>::reduceDim(int dim) {
-    }
-  }
+  int stopDim = (excludeDim == -1) ? dims : excludeDim;
+  int nidx = -1;
```
can you make the names a little more descriptive?

maybe `nidx -> newIndex`, `cidx -> oldIndex`, `rVal -> remapDim` say?

wickedfoo reviewed (approved):
Looks fine to me (I'm the original author of this from way back when). The only change I'd make is to make the index variable names a little more descriptive.

Torch has always lacked good testing of contiguous vs. non-contiguous element processing, especially in some of the more exotic configurations.

ezyang commented:
@mruberry, would it be possible to mainline some of the extra tests you did into the test suite? Every extra bit helps, especially for tricky cases like this.

mruberry commented:
@ezyang 

I would really like to add some tests. What is your recommendation for how to do that?

I don't think there's a way to test collapseDims() directly from Python, so would your recommendation be to extend the C++ tests in ATen/test? That may be appropriate since unless someone changes the actual function they likely don't need to run the tests for it. 

Let me know the extensible thing to do and I'll write the tests.

ezyang commented:
C++ tests sound like just the ticket. Don't forget to add `AT_API` to the function in that case, so the Windows build doesn't fail.

ezyang commented:
Also, I can merge this PR if you want to work on the tests in a separate PR. Let me know what you prefer.

mruberry commented:
Thanks for the offer; I'd rather put the tests in, first.

Good thing you mentioned AT_API! I'll create a new .cpp file in ATen/test. I was planning on patterning it off the files that use catch, like integer_divider_test.cu and native_test.cpp, but none of the files seem to use AT_API. Is there an example of a cpp test that does use AT_API?

ezyang commented:
Just grep for it; if you want to access a function from a test executable, you have to label it with `AT_API` (this handles Windows which has different visibility semantics.)

mruberry committed:
Adds C++ tests, validates input, updates names for readability

mruberry commented:
I think all necessary changes have been made. The AT_API macro now appears before the TensorInfo class in ATen. Although I did run these tests on the function in THC, too, this submitted version does not test the THC version because I did not want to import AT_API to THC. The functions are identical, however, so that testing is likely not needed.

The additional test file may be useful in the future for testing other tensor behavior related to pointwise apply, too.

mruberry commented:
The Windows failure appears to be incorrect use of AT_API:

23:30:07 [396/443] Building CXX object src\ATen\test\CMakeFiles\apply_test.dir\apply_test.cpp.obj
23:30:07 ..\src\ATen/cuda/detail/TensorInfo.cuh(63): error C2491: 'at::cuda::detail::TensorInfo<T,IndexType>::TensorInfo': definition of dllimport function not allowed
23:30:07 ..\src\ATen/cuda/detail/TensorInfo.cuh(70): error C2491: 'at::cuda::detail::TensorInfo<T,IndexType>::reduceDim': definition of dllimport function not allowed
23:30:07 ..\src\ATen/cuda/detail/TensorInfo.cuh(139): error C2491: 'at::cuda::detail::TensorInfo<T,IndexType>::collapseDims': definition of dllimport function not allowed

Unfortunately, I don't know anything about Windows development and don't have a Windows machine readily available. Is there someone familiar with Windows for who would immediately know the appropriate fix here? 

(Looking at the Linux failure now.)

mruberry commented:
The linux-xenial-cuda8-cudnn6-py3 and linux-xenial-cuda9-cudnn7-py2 failures are both caused when trying to trigger the assertion using REQUIRE_THROWS(ti.collapseDims(5)). 

I am not totally sure why this is happening. Would you recommend removing this test? The assertion replaces several assertions in the original and is comparable to another assert in reduceDim(), so that seems appropriate to keep.

mruberry committed:
Removes invalid test

ezyang commented:
Two things:

1. An "assert" should be reserved for things which are truly impossible to trigger unless an ATen developer has messed up their code. In this case, collapseDims is semi-public, so it's not appropriate for it to use asserts; it should use the `AT_CHECK` macro instead.

2. Even for asserts, you should use the `AT_ASSERT` macro anyway, which raises an exception.

There are a few places we are using `assert.h` but they are wrong and should be converted to use our assert macro instead.

mruberry committed:
stashing to merge AT_CHECK macro

mruberry committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into collapse_dims

mruberry committed:
Updates asserts, removes tests on Windows

mruberry commented:
@ezyang 

Thanks for pointing that out. I not only updated that in this function and put the removed test back in, but also adjusted the other asserts in the ATen version of the file.

I removed the Windows code and disabled the test on Windows. They can be easily enabled on Windows in the future if there's a plan for exporting class templates. (I believe Windows requires an instantiation to be exported, and it's not clear in the current code base where these exports should live.)
